posts/promql-guide.md

```diff
@@ -0,0 +1,159 @@
+# PromQL for Humans
+
+PromQL is a built in query-language made for Prometheus. Here at Timber [we've found Prometheus to be awesome](https://timber.io/blog/prometheus-the-good-the-bad-and-the-ugly/), but PromQL difficult to wrap our heads around. This is our attempt to change that.
+
+## Basics
+
+### Instant Vectors
+
+_Only Instant Vectors can be graphed._
+
+`http_requests_total`
+
+![](./images/promql-guide/http_requests_total.png)
+
+This gives us all the http requests, but we've got 2 issues.
+1. There are too many data points to decipher what's going on.
+2. You'll notice that `http_requests_total` only goes up, because it's a [counter](https://prometheus.io/docs/concepts/metric_types/#counter). These are common in Prometheus, but not useful to graph.
+
+I'll show you how to approach both.
+
+It's easy to filter by label.
+
+`http_requests_total{job="prometheus", code="200"}`
+
+`!=` - not equal
+
+![](./images/promql-guide/filter-by-label.png)
+
+You can check a substring using regex matching.
+`http_requests_total{status_code=~"2.*"}`
+
+`!~` - does not regex match
+
+![](./images/promql-guide/substring.png)
+
+If you're interested in learning more, [here](https://docs.python.org/3/library/re.html) are the docs on Regex.
+
+### Range Vectors
+
+_Contain data going back in time._
+
+Recall: Only Instant Vectors can be graphed. You'll soon be able to see how to visualize Range Vectors using functions.
+
+`http_requests_total[5m]`
+
+You can also use (s, m, h, d, w, y) to represent (seconds, minutes, hours, ...) respectively.
+
+## Important functions
+
+### For Range Vectors
+
+_You'll notice that we're able to graph all these functions. Since only Instant Vectors can be graphed, they take a Range Vector as a parameter and return a Instant Vector._
+
+`rate(http_requests_total[5m])` - increase of `http_requests_total` averaged over the last 5 minutes. Use this when alerting because it creates a smooth graph.
```

Comment:
Why do we want a smooth graph. Just a sentence or phrase e.g. Spikey graphs can cause alert overload, fatigue, and bad times for all due to repeatedly triggering thresholds.